app/helpers/graph_utils.rb

```diff
@@ -0,0 +1,151 @@
+class GraphUtils
+
+  def self.assign_graders(submissions, graders, weights, prohibitions)
+    # Given a list of submissions, a list of graders,
+    #   a mapping of grader IDs to weights of how much they should work,
+    #   and a mapping of grader IDs to lists of student IDs that the grader should not grade,
+    #   assign each submission to a grader, if possible.
+    # Runs in time O(V^3) = O((submissions + graders) ^ 3) = O(submissions ^ 3)
+    #   when graders < submissions, which is basically always.
+    # We could implement the highest-level rule, so that it runs in time O(V^2 E^1/2),
+    #   but E = submissions * graders, so we'd get O(submissions ^ 2.5) at best,
+    #   and the inefficiencies due to lack of decent data structures means we
+    #   wouldn't gain much after all the bookkeeping is done.
+    total_weight = weights.values.sum
+    sub_to_n = {}
+    grader_to_n = {}
+    n_to_grader = {}
+    n = 1
+    submissions.each do |s|
+      sub_to_n[s.id] = n
+      n += 1
+    end
+    graders.each do |g|
+      grader_to_n[g.id] = n
+      n_to_grader[n] = g
+      n += 1
+    end
+    c = {}
+    c[0] = {}
+    c[n] = {}
+    # Connect 1 unit of flow from the source to each submission
+    submissions.each do |s|
+      c[0][sub_to_n[s.id]] = 1
+    end
+    # Connect an appripriate weight of work from each grader to the sink
+    graders.each do |g|
+      c[grader_to_n[g.id]] = {}
+      c[grader_to_n[g.id]][n] = ((weights[g.id].to_f / total_weight) * submissions.size).ceil
```

Comment:
Instead of ceiling'ing every grader's capacity, 
* start by recording the fractional values.  
* Then tally up `allocated_capacity = sum of floor'ed capacities`, to see how many submissions are definitely accounted for.  
* In decreasing sorted order by grader's capacity, ceiling _just that grader's capacity_, and increment `allocated_capacity` until it reaches `#submissions`.
* At that point, floor everyone else's capacity.

This should ensure that the actual capacity to the sink is exactly `#submissions`, without any slack left over, which should ensure an equitable and evenly balanced workload for everyone.